### DISCUSSION TITLE
JPH_ASSERT fails on convex hull shape, but ConvexHullBuilder succeeds.

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#991] [Q&A]
I get a Success result from ConvexHullBuilder, but as soon as the first step of the simulation occurs, I get a JPH_ASSERT failure on line 104 in EPAPenetrationDepth.h.
(Yes, I know that ConvexHullBuilder is not necessary here, but I added this to validate that the hull was correctly being created).

What could be going on here? These are not complex hulls, they are akin to simplified cyndrilical platforms.

`

		// Don't supply a zero ioV, we only want to get points on the hull of the Minkowsky sum and not internal points
		JPH_ASSERT(!ioV.IsNearZero());

`

`

	JPH::ConvexHullBuilder builder(vertexList);
	const char *error = nullptr;
	float tolerance = 1.0e-3f;
	JPH::ConvexHullBuilder::EResult result = builder.Initialize(INT_MAX, tolerance, error);
	if (result != JPH::ConvexHullBuilder::EResult::Success && result != JPH::ConvexHullBuilder::EResult::MaxVerticesReached)
	{
		JPH::Trace("Iteration %d: Failed to initialize from positions: %s", 1, error);
		JPH_ASSERT(false);
		return;
	}

	JPH::BodyCreationSettings bcs(new JPH::ConvexHullShapeSettings(vertexList), mat.GetTranslation(), mat.GetQuaternion(), JPH::EMotionType::Dynamic, Layers::MOVING);
	bcs.mUserData = (JPH::uint64)object;
	bcs.mRestitution = 0;
	bcs.mAngularDamping = 0;
	bcs.mFriction = 0;
	bcs.mMaxAngularVelocity = 0;

	object->collisionObjectID = bi.CreateAndAddBody(bcs, JPH::EActivation::Activate);

`

**Comment by jrouwe** [comment]
Usually this means that there is a degenerate triangle in the system, but with a convex hull this should not be possible.

Do you have the full callstack of the assert for me?

If it comes from where I think it should come from then it should have passed these lines:

https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Jolt/Physics/Collision/Shape/ConvexShape.cpp#L74-L75

So the value should have changed from zero to (1, 0, 0).

**Comment by SSNTails** [comment]
Apologies for the screenshot, I don't know how to dump the call stack as a string from Visual Studio:

![image](https://github.com/jrouwe/JoltPhysics/assets/159929665/d387b20b-b180-43d8-8bf0-295ba69605c3)

Did a little bit more digging, and it's an object made of two cyndrilical meshes. Granted, it is created from 1512 vertices, but I don't think that would make a difference (I used to use GImpact with Bullet to load arbitrary kinematic triangle meshes and I'm just trying to get things up and running - I figured making them convex hulls for now would be sufficient).

**Comment by jrouwe** [comment]
Dumping a callstack is as simple as clicking on the callstack window, Ctrl+A, Ctrl+C.

In any case the issue is not with your convex hull, it is with the mesh that it is hitting. I'm not sure which version of Jolt you're using but I [recently](https://github.com/jrouwe/JoltPhysics/commit/cc49e07d7c4eb854a60cb21ea0abdfcc93511410) added an extra check for degenerate triangles there. Basically if you have a triangle that is degenerate or one that becomes degenerate after quantization you'll trigger this assert. This new code will remove them automatically.

**Reply by SSNTails** [reply]
I'm using the latest release, 4.0.2. I'll have to pull the master branch to get these changes, looks like.

It's highly likely it's a triangle issue, as sliding the player over one particular area near there causes Jolt to flip out. I was finding all of the error checking quite unnerving at first, but I'm getting used to it! 😅
